File: 2nd year/MPP/project-mpp/java-project/TriathlonServer/src/main/java/triathlon/server/SuperService.java
package triathlon.server;

import triathlon.model.Athlete;
import triathlon.model.Game;
import triathlon.model.Referee;
import triathlon.model.Result;
import triathlon.persistence.AthleteRepository;
import triathlon.persistence.GameRepository;
import triathlon.persistence.RefereeRepository;
import triathlon.persistence.ResultRepository;
import triathlon.services.ITriathlonObserver;
import triathlon.services.ITriathlonServices;
import triathlon.services.TriathlonException;

import java.rmi.RemoteException;
import java.util.Collection;
import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.stream.Collectors;
import java.util.stream.StreamSupport;

public class SuperService implements ITriathlonServices {
    private final GameRepository gameRepository;
    private final ResultRepository resultRepository;
    private final RefereeRepository refereeRepository;
    private final AthleteRepository athleteRepository;
    private final Map<String, ITriathlonObserver> loggedClients;

    public SuperService(GameRepository gameRepository, ResultRepository resultRepository,
                        RefereeRepository refereeRepository, AthleteRepository athleteRepository) {
        this.gameRepository = gameRepository;
        this.resultRepository = resultRepository;
        this.refereeRepository = refereeRepository;
        this.athleteRepository = athleteRepository;
        loggedClients = new ConcurrentHashMap<>();
    }

    @Override
    public synchronized Collection<Athlete> getAthletes() {
        return StreamSupport
                .stream(athleteRepository.findAll().spliterator(), false)
                .collect(Collectors.toList());
    }

    @Override
    public synchronized Map<String, Float> getAthletesTotalPoints() {
        Map<String, Float> athleteResultMap = new HashMap<>();
        for (Result result : resultRepository.findAll()) {
            Athlete athlete = athleteRepository.findOne(result.getAthlete().getId());
            athleteResultMap.compute(athlete.getName(), (k, v) -> (v == null ? 0 : v) + result.getValue());
        }
        return athleteResultMap;
    }

    @Override
    public synchronized Game getGameById(Long id) {
        return gameRepository.findOne(id);
    }

    @Override
    public synchronized Referee authenticate(Referee referee, ITriathlonObserver client) throws TriathlonException {
        for (Referee ref : refereeRepository.findAll()) {
            if (ref.getUsername().equals(referee.getUsername()) &&
                    ref.getPassword().equals(referee.getPassword())) {
                if (loggedClients.containsKey(ref.getUsername()))
                    throw new TriathlonException("User already logged in!");
                loggedClients.put(ref.getUsername(), client);
                return ref;
            }
        }
        return null;
    }

    @Override
    public synchronized void logout(Referee referee, ITriathlonObserver client) throws TriathlonException {
        System.out.println("Logout: " + referee.getUsername());
        ITriathlonObserver localClient = loggedClients.remove(referee.getUsername());
        if (localClient == null) {
            throw new TriathlonException("Referee " + referee.getId() + " is not logged in.");
        }
    }

    @Override
    public synchronized Collection<Result> getResultsForGame(Long gameId) {
        return StreamSupport
                .stream(resultRepository.findAll().spliterator(), false)
                .filter(result -> result.getGame().getId().equals(gameId))
                .collect(Collectors.toList());
    }

    @Override
    public synchronized void setResult(Athlete athlete, Game game, Float value) {
        Result result = StreamSupport.stream(resultRepository.findAll().spliterator(), false)
                .filter(r -> r.getAthlete().getId().equals(athlete.getId()) && r.getGame().getId().equals(game.getId()))
                .findFirst()
                .orElse(null);

        if (result != null) {
            result.setValue(value);
            resultRepository.update(result);
        } else {
            resultRepository.save(new Result(game, athlete, value));
        }
        notifyUpdateResult(result);
    }

    private void notifyUpdateResult(Result result) {
        ExecutorService executor = Executors.newFixedThreadPool(5);
        for (String username : loggedClients.keySet()) {
            ITriathlonObserver client = loggedClients.get(username);
            if (client != null) {
                executor.execute(() -> {
                    try {
                        System.out.println("Notifying [" + username + "] to update its info.");
                        client.pointsChanged(result);
                    } catch (TriathlonException | RemoteException e) {
                        System.err.println("Error notifying operator " + e);
                    }
                });
            }
        }
        executor.shutdown();
    }
}

File: 2nd year/MPP/project-mpp/csharp-project/server/SuperService.cs
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using model;
using persistence;
using services;

namespace server
{
    public class SuperService : ITriathlonServices
    {
        private readonly IGameRepository _gameRepository;
        private readonly IResultRepository _resultRepository;
        private readonly IAthleteRepository _athleteRepository;
        private readonly IRefereeRepository _refereeRepository;
        private readonly IDictionary<string, ITriathlonObserver> _loggedClients;

        public SuperService(IGameRepository gameRepository, IResultRepository resultRepository,
            IAthleteRepository athleteRepository, IRefereeRepository refereeRepository)
        {
            _gameRepository = gameRepository;
            _resultRepository = resultRepository;
            _athleteRepository = athleteRepository;
            _refereeRepository = refereeRepository;
            _loggedClients = new Dictionary<string, ITriathlonObserver>();
        }

        public Referee Authenticate(Referee inputReferee, ITriathlonObserver client)
        {
            foreach (var referee in _refereeRepository.FindAll())
            {
                if (referee.Username.Equals(inputReferee.Username) && referee.Password.Equals(inputReferee.Password))
                {
                    if (_loggedClients.ContainsKey(referee.Username))
                    {
                        throw new TriathlonException("User already logged in.");
                    }

                    _loggedClients[referee.Username] = client;
                    return referee;
                }
            }

            return null;
        }

        public IList<Athlete> GetAthletes()
        {
            return _athleteRepository.FindAll().ToList();
        }

        public Dictionary<string, double> GetAthletesTotalPoints()
        {
            Dictionary<string, double> athleteResultMap = new Dictionary<string, double>();
            foreach (var result in _resultRepository.FindAll())
            {
                var athlete = _athleteRepository.FindOne(result.Athlete.Id);
                if (athleteResultMap.ContainsKey(athlete.Name))
                {
                    athleteResultMap[athlete.Name] += result.Value;
                }
                else
                {
                    athleteResultMap.Add(athlete.Name, result.Value);
                }
            }

            return athleteResultMap;
        }

        public Game GetGameById(long id)
        {
            return _gameRepository.FindOne(id);
        }

        public IList<Result> GetResultsForGame(long gameId)
        {
            return _resultRepository.FindAll()
                .Where(result => result.Game.Id == gameId)
                .Select(result => new Result
                {
                    Athlete = _athleteRepository.FindOne(result.Athlete.Id),
                    Game = result.Game,
                    Id = result.Id,
                    Value = result.Value
                }).ToList();
        }

        public void Logout(Referee referee, ITriathlonObserver client)
        {
            var localClient = _loggedClients[referee.Username];
            if (localClient == null)
            {
                throw new TriathlonException("Referee " + referee.Username + " is not logged in.");
            }

            _loggedClients.Remove(referee.Username);
        }

        public void SetResult(Athlete athlete, Game game, double value)
        {
            var result = (from r in _resultRepository.FindAll()
                where r.Athlete.Id == athlete.Id && r.Game.Id == game.Id
                select new Result
                {
                    Athlete = athlete,
                    Game = game,
                    Value = value,
                    Id = r.Id
                }).FirstOrDefault();
            if (result != null)
            {
                result.Value = value;
                _resultRepository.Update(result);
            }
            else
            {
                result = new Result()
                {
                    Game = game,
                    Athlete = athlete,
                    Value = value
                };
                _resultRepository.Save(result);
            }

            NotifyUpdateResult();
        }

        private void NotifyUpdateResult()
        {
            foreach (var client in _loggedClients)
            {
                Task.Run(() => client.Value.PointsChanged());
            }
        }
    }
}
